contracts/votingMachines/GenesisProtocolLogic.sol

```diff
@@ -0,0 +1,660 @@
+        uint executionBar = totalReputation * params.queuedVoteRequiredPercentage/100;
+        ExecutionState executionState = ExecutionState.None;
+        uint averageBoostDownstakes;
+
+        if (proposal.votes[proposal.winningVote] > executionBar) {
+         // someone crossed the absolute vote execution bar.
+            if (proposal.state == ProposalState.Queued) {
+                executionState = ExecutionState.QueueBarCrossed;
+            } else if (proposal.state == ProposalState.PreBoosted) {
+                executionState = ExecutionState.PreBoostedBarCrossed;
+            } else {
+                executionState = ExecutionState.BoostedBarCrossed;
+            }
+            proposal.state = ProposalState.Executed;
+         } else {
+            if (proposal.state == ProposalState.Queued) {
+                // solium-disable-next-line security/no-block-members
+                if ((now - proposal.times[0]) >= params.queuedVotePeriodLimit) {
+                    proposal.state = ProposalState.ExpiredInQueue;
+                    proposal.winningVote = NO;
+                    executionState = ExecutionState.QueueTimeOut;
+                 } else if ( shouldBoost(_proposalId)) {
+                    //change proposal mode to PreBoosted mode.
+                    proposal.state = ProposalState.PreBoosted;
+                    // solium-disable-next-line security/no-block-members
+                    proposal.times[2] = now;
+                    proposal.confidenceThreshold = threshold(proposal.paramsHash,proposal.organizationId);
+                  }
+               }
+
+            if (proposal.state == ProposalState.PreBoosted) {
+              // solium-disable-next-line security/no-block-members
+                if ((now - proposal.times[2]) >= params.preBoostedVotePeriodLimit) {
+                    if (shouldBoost(_proposalId)) {
+                     //change proposal mode to Boosted mode.
+                        proposal.state = ProposalState.Boosted;
+                       // solium-disable-next-line security/no-block-members
+                        proposal.times[1] = now;
+                        orgBoostedProposalsCnt[proposal.organizationId]++;
+                       //add a value to average -> average = average + ((value - average) / nbValues)
+                        averageBoostDownstakes = averagesBoostDownstakes[proposal.organizationId];
+                        averagesBoostDownstakes[proposal.organizationId] = uint256(int256(averageBoostDownstakes) + ((int216(proposal.stakes[NO])-int216(averageBoostDownstakes)).toReal().div(int216(orgBoostedProposalsCnt[proposal.organizationId]).toReal())).fromReal());
+                 }
+               } else { //check the Confidence level is stable
+                    if (_score(_proposalId) <= proposal.confidenceThreshold) {
+                        proposal.state = ProposalState.Queued;
```

Comment:
At this point, if going back to queued, should we rerun the logic for `if (proposal.state == ProposalState.Queued)`?

Reply:
why?

Reply:
because at this point if `if ((now - proposal.times[0]) >= params.queuedVotePeriodLimit)` then the proposal  should timeout, else if `shouldBoost(_proposalId)`  then it should restart being preboosted.

Reply:
if it is timeout after getting back to the q ..it is the same as it timed out as if it was not at the q at all....so I do not see a need to spend some more gas on that .
and also it cannot get back to preboost immediately (otherwise it will loop forever)

Reply:
the only reason I see for doing it is to expedite the transition to a timeout that will have (in this case) already occurred.

Reply:
see my comments above

Reply:
The only reason for understaking my suggestion is the expidite the transaction to other stages that might be immediately apparent.

Reply:
Sorry @orenyodfat I edited my comment after you replied to it, not realizing you had replied.  Here it is again:

I don't think you'll go into an infinite loop because you are resetting proposal.times[2] and confidenceThreshold (there would be something wrong with the logic otherwise)

Reply:
As I wrote ,I don't see why it is needed.

Reply:
It is needed if we care that switching to "queued" may be completely incorrect.

Reply:
yes,  if `((now - proposal.times[0]) >= params.queuedVotePeriodLimit)` then the proposal should be timed-out, not set to queued, else if `shouldBoost(_proposalId)` then it should restart as  preboosted, not queued.

Reply:
"no need to check its timeout"?